module/zfs/zfs_ioctl.c

```diff
@@ -2929,6 +2929,36 @@ zfs_set_targetip_posthook(const char *name, char *targetip, char *curtargetip)
 		nvlist_free(props);
 	}
 }
+
+/*
+ * zfs_set_replicaid_hook is used to set replica_id from given properties.
+ * inputs:
+ * - name: name of the dataset
+ * - nvl: property list
+ */
+int
+zfs_set_replicaid_hook(const char *name, nvlist_t *nvl)
+{
+	char *replica_id = NULL, *end = NULL;
+	nvlist_t *props;
+	int error = 0;
+
+	if (nvlist_lookup_string(nvl, ZFS_PROP_ZVOL_REPLICA_ID,
+	    &replica_id) == 0) {
+		if (strtoul(replica_id, &end, 10) == 0) {
+			error = SET_ERROR(EINVAL);
+		} else if (*end != '\0') {
+			error = SET_ERROR(EINVAL);
+		} else {
+			nvlist_alloc(&props, NV_UNIQUE_NAME, 0);
+			nvlist_add_string(props, ZFS_PROP_ZVOL_REPLICA_ID,
+			    replica_id);
+			error = uzfs_zvol_create_cb(name, props);
```

Comment:
we should not call uzfs_zvol_create_cb here. we should call it from targetip_prehook(which is already calling this)

Reply:
During upgrade, dataset will have targetip set but not replica_id. In that case, to block startup of zinfo, we need to have this.